src/gsi/cplr_get_fv3_regional_ensperts.f90

```diff
@@ -430,112 +543,159 @@ subroutine general_read_fv3_regional(this,fv3_filenameginput,g_ps,g_u,g_v,g_tv,g
+    real(r_kind),parameter:: r10   = 10.0_r_kind
+    real(r_kind),parameter:: r100  = 100.0_r_kind
+   !
+!   Declare local variables
+    
+    integer(i_kind):: i,j,k,kp
+ 
+    integer(i_kind) iderivative
+
+    
+    logical ice
 
-      character(len=24),parameter :: myname_ = 'general_read_fv3_regional'
+    character(len=24),parameter :: myname_ = 'general_read_fv3_regional'
+    type(gsi_bundle) :: gsibundle_fv3lam_ens_dynvar_nouv
+    type(gsi_bundle) :: gsibundle_fv3lam_ens_tracer_nouv
+    type(gsi_grid):: grid_ens
 
-      character(len=:),allocatable :: grid_spec !='fv3_grid_spec'            
-      character(len=:),allocatable :: ak_bk     !='fv3_akbk'
-      character(len=:),allocatable :: dynvars   !='fv3_dynvars'
-      character(len=:),allocatable :: tracers   !='fv3_tracer'
-      character(len=:),allocatable :: sfcdata   !='fv3_sfcdata'
-      character(len=:),allocatable :: couplerres!='coupler.res'
-      
-      associate( this => this ) ! eliminates warning for unused dummy argument needed for binding
-      end associate
+    character(len=:),allocatable :: grid_spec !='fv3_grid_spec'            
+    character(len=:),allocatable :: ak_bk     !='fv3_akbk'
+    character(len=:),allocatable :: dynvars   !='fv3_dynvars'
+    character(len=:),allocatable :: tracers   !='fv3_tracer'
+    character(len=:),allocatable :: sfcdata   !='fv3_sfcdata'
+    character(len=:),allocatable :: couplerres!='coupler.res'
+    integer (i_kind) ier,istatus
+
+    
+    associate( this => this ) ! eliminates warning for unused dummy argument needed for binding
+    end associate
 
 
 
+    call gsi_gridcreate(grid_ens,grd_ens%lat2,grd_ens%lon2,grd_ens%nsig)
     grid_spec=fv3_filenameginput%grid_spec
     ak_bk=fv3_filenameginput%ak_bk
     dynvars=fv3_filenameginput%dynvars
     tracers=fv3_filenameginput%tracers
     sfcdata=fv3_filenameginput%sfcdata
     couplerres=fv3_filenameginput%couplerres
 
-      
 
-!cltthinktobe  should be contained in variable like grd_ens
+     
+     
+    if (allocated(fv3lam_ens_io_dynmetvars2d_nouv) ) then   
+       call gsi_bundlecreate(gsibundle_fv3lam_ens_dynvar_nouv,grid_ens,'gsibundle_fv3lam_ens_dynvar_nouv', istatus,&
+                names2d=fv3lam_ens_io_dynmetvars2d_nouv,names3d=fv3lam_ens_io_dynmetvars3d_nouv)
+    else
+       call gsi_bundlecreate(gsibundle_fv3lam_ens_dynvar_nouv,grid_ens,'gsibundle_fv3lam_ens_dynvar_nouv',istatus, &
+                names3d=fv3lam_ens_io_dynmetvars3d_nouv)
+    endif
+
+     
+    if (allocated(fv3lam_ens_io_tracermetvars2d_nouv) ) then   
+       call gsi_bundlecreate(gsibundle_fv3lam_ens_tracer_nouv,grid_ens,'gsibundle_fv3lam_ens_tracer_nouv', istatus,&
+                names2d=fv3lam_ens_io_tracermetvars2d_nouv,names3d=fv3lam_ens_io_tracermetvars3d_nouv)
+    else
+       call gsi_bundlecreate(gsibundle_fv3lam_ens_tracer_nouv,grid_ens,'gsibundle_fv3lam_ens_tracer_nouv',istatus, &
+                names3d=fv3lam_ens_io_tracermetvars3d_nouv)
+    endif
```

Comment:
Line 634 and Line 643:  I don't understand why the 2D/3D bundle creation is switched by
   if (allocated(fv3lam_ens_io_dynmetvars2d_nouv) )
and
   if (allocated(fv3lam_ens_io_tracermetvars2d_nouv) )

(1) Should them be "if (ndynvario2d > 0 )" and "if (ntracerio2d > 0 )" instead?
(2) Does it mean that 2D and 3D variables cannot exist at the same time?

Reply:
Currently, if ,for example, ndyvar2d>0 , then fv3lam_ens_io_dynmetvars2d_nouv would allocated.  I chose to use the current form rather to use ,say, ndyvar2d because the former is more "essential" , while ndyvar2d might change as a local variable in this subroutine.  When the character allocatable arrays for 2d variable is allocated, that mean 2d variables are to be dealt with in the IO and hence would be used in the gsi_bundlecreate along with the 3d variable character array , otherwise, only the 3d variable arrays would be used in the gsi_bundlecreate.